### ISSUE TITLE
Handling Out-of-memory situations when template loading

### ISSUE THREAD

**Issue by yevhenmazur** [OpenOrienteering/mapper#1273]
### Steps to reproduce
1. Make visible big raster template

### Behavior in dev v20190622.5
Application crashed

### Behavior in 0.8.4 
The message appears: 'Failed to load template ... Not enough free memory'

### Configuration
Operating System: Android

**Comment by dg0yt** [comment]
Out of memory handling for templates is unchanged. We catch `std::bad_alloc` in general:

https://github.com/OpenOrienteering/mapper/blob/9adf5c7053de8dafeb03219815d1c89efe40eba4/src/templates/template.cpp#L570-L574

and `QImage::isNull` in particular:

https://github.com/OpenOrienteering/mapper/blob/9adf5c7053de8dafeb03219815d1c89efe40eba4/src/templates/template_image.cpp#L155-L160

So what error message did you see exactly in 0.8.4?

What has changed in dev is Mapper's environment: Qt 5.12, a newer Android NDK, a new target API level. The latter may give another reason for the app "crashing": Android would close the app if some of the target API level requirements are broken by our app.

So what is the Android version of your device?

**Comment by yevhenmazur** [comment]
The error message in 0.8.4

![Screenshot_2019-06-24-22-05-24-713_org openorienteering mapper](https://user-images.githubusercontent.com/10380431/60045183-dc98b180-96cc-11e9-8471-46c172574114.png)

Android version: 9 PKQ1.181121.001

**Comment by krticka** [comment]
I think that such big raster can consume at least 3GB of memory. Maybe this is the problem?

**Comment by dg0yt** [comment]
Yes, this image will need almost 4 GB of RAM (with 32 bits per pixel), which is the maximum a 32 bit system can handle in total. However, it should not crash.
- Since v0.8.4 (Qt 5.6), `QImage` learned to [handle more than 2 GB image data](https://code.qt.io/cgit/qt/qtbase.git/commit/src/gui/image/qimage.cpp?id=4f88475a962975ca45994cff9add350344fce4f9). This is where different behaviour starts now.
- The crash might be related to memory overcommitment: The operating system may grant more virtual memory to the app than can be accessed physically later (even with swapping memory between RAM and flash storage), eventually forcing the system to terminate the app. This is hard to avoid in advance. 
- Assuming you were using the 32 bit package: Maybe the 64 bit package works on your device and does not crash?

Maybe we can avoid such situations by limiting the amount of memory we allow for template images with respect to the amount of RAM in the system. 
In the long term, we shouldn't always load whole template images at once, but that's a different story. (GDAL raster support has interfaces to deal with that.)

**Comment by dg0yt** [comment]
Android links:
- [Monitor available memory and memory usage](https://developer.android.com/topic/performance/memory#monitor)
- [ActivityManager.MemoryInfo](https://developer.android.com/reference/android/app/ActivityManager.MemoryInfo)
